### PR TITLE
⚠️ Adapt clusterctl to webhook deployed with managers

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
After https://github.com/kubernetes-sigs/cluster-api/pull/3985 Cluster API and all the providers are going to deploy web hook with managers.

This makes it possible to cleanup some code in clusterctl and to simplify a few interfaces in the clusterctl library

**Which issue(s) this PR fixes**:
Rif https://github.com/kubernetes-sigs/cluster-api/issues/4056

### CODE REVIEW

**Comment by stmcginnis** [comment]
Changes look good to me, but are the failures in `pull-cluster-api-apidiff-main` and issue?

Very nice! >> "+75 −660"

**Comment by vincepri** [comment]
```
sigs.k8s.io/cluster-api/cmd/clusterctl/client
  Incompatible changes:
  - sigs.k8s.io/cluster-api/cmd/clusterctl/client.Components.InstanceObjs: removed
  - sigs.k8s.io/cluster-api/cmd/clusterctl/client.Components.Objs: added
  - sigs.k8s.io/cluster-api/cmd/clusterctl/client.Components.SharedObjs: removed
sigs.k8s.io/cluster-api/cmd/clusterctl/api/v1alpha3
  Incompatible changes:
  - ClusterctlResourceLifecyleLabelName: removed
  - ResourceLifecycle: removed
  - ResourceLifecycleShared: removed
sigs.k8s.io/cluster-api/cmd/clusterctl/client/repository
  Incompatible changes:
  - (*components).InstanceObjs: removed
  - (*components).SharedObjs: removed
  - Components.InstanceObjs: removed
  - Components.Objs: added
  - Components.SharedObjs: removed
  - WebhookNamespaceName: removed
  Compatible changes:
  - (*components).Objs: added
+ EXIT_VALUE=1
+ set +o xtrace
```

Should we mark this as breaking?

**Comment by cpanato** [comment]
What is the impact on the providers? what we should do when updating the dependencies that will have this change here?

**Comment by vincepri** [comment]
@cpanato See https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md#required-kustomize-changes-to-have-a-single-manager-watching-all-namespaces-and-answer-to-webhook-calls

**Comment by fabriziopandini** [comment]
@cpanato impacts on the provider are already documented in https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md#multi-tenancy; this PR cleanups clusterctl accordingly

**Comment by cpanato** [comment]
thanks for the clarification @fabriziopandini @vincepri 

LGTM

File: cmd/clusterctl/client/cluster/components.go [hunk 1]
```diff
@@ -170,16 +167,17 @@ func (p *providerComponents) Delete(options DeleteOptions) error {
...
+		// on that for deleting only the global resources belonging the the instance we are processing.
+		// NOTE: namespace and CRD are special case managed above; webhook instead goes hand by hand with the controller they
+		// should always be deleted.
+		isWebhook := obj.GroupVersionKind().Kind == validatingWebhookConfigurationKind || obj.GroupVersionKind().Kind == mutatingWebhookConfigurationKind
+		if util.IsClusterResource(obj.GetKind()) &&
+			!(isNamespace || isCrd || isWebhook) {
+			if !strings.HasPrefix(obj.GetName(), instanceNamespacePrefix) {
+				continue
 			}
 		}
```
Comment (sbueringer) [inline cmd/clusterctl/client/cluster/components.go:182]:
```suggestion
		if util.IsClusterResource(obj.GetKind()) &&
			!isNamespace && !isCrd && !isWebhook &&
			!strings.HasPrefix(obj.GetName(), instanceNamespacePrefix) {
			continue
		}
```
Maybe that's easier to comprehend (hope it's equivalent)

**Review by sbueringer — COMMENTED** [review]
code looks plausible and good to me, but I'm definitely missing too much context

File: cmd/clusterctl/client/cluster/components.go (hunk 1 above)
Comment (fabriziopandini) [reply cmd/clusterctl/client/cluster/components.go:182]: done!

**Review by vincepri — APPROVED** [review]
/approve
/lgtm

**Pull request merged**